### ISSUE TITLE
Add support of Mathjax

### ISSUE THREAD

**Issue by kelson42** [openzim/mwoffliner#1371]
https://farm.openzim.org/recipes/proofwiki Does not render properly

**Comment by MananJethwani** [comment]
@kelson42 it doesn't render properely because not all modules required in the page are being downloaded api.php is probably depriciated.

**Comment by kelson42** [comment]
@MananJethwani Its not deprecated as far as I know, wee need to download the mathjax module. Here (again) it should have happened automatically as this is for surce a resourceLoader dependence of the page.

**Comment by kelson42** [comment]
One problem is that Mediawiki does not list the resourceLoader properly in the deps, see https://github.com/openzim/mwoffliner/issues/1391#issuecomment-773242914

**Comment by kelson42** [comment]
@Krinkle I have finally create the upstream ticket at https://phabricator.wikimedia.org/T278061. Would you be able to fix it yourself?

**Comment by trappedinspacetime** [comment]
Is it possible to inject that version of `MathJax`?

**Comment by Havael** [comment]
Has there been any progress?

**Comment by kelson42** [comment]
@Havael Unfortunetaly no, I have checked again with latest version and we have still fully broken  Math formulas. See https://browse.library.kiwix.org/viewer#proofwiki_en_all_maxi_2025-06/Main_Page
@benoit74 The upstream issue at Phabricator has been closed. I'm really blind here where is the source of the problem. Dow we have here too an issue related to the list of modules or gadgets... Or an other kind?

**Comment by Markus-Rost** [comment]
Mathjax is a JS module and we currently don't load JS at all (see #2310). Additionally the module `ext.mathJax` isn't even listed by ActionParse (see [T161278](https://phabricator.wikimedia.org/T161278)), so even if we were to load JS modules we would still be missing MathJax.

MathML is supported by all browsers since 2023, so returning `<math>` elements instead of plain text would be the ideal solution. MathJax can still be used on the `<math>` elements for support of older browser.

Looks like ProofWiki is using a custom extension for more input styles so just switching to [Extension:Math](https://www.mediawiki.org/wiki/Extension:Math#MathJax) wouldn't work, but maybe they can update their own extension to do some server-side parsing into `<math>` elements.

**Comment by Jaifroid** [comment]
> MathML is supported by all browsers since 2023

That's pretty recent for Kiwix users, and especially those in Global South, with older devices, etc.

**Comment by kelson42** [comment]
We have to find a solution here in 2.0.0.

**Comment by kelson42** [comment]
@benoit74 It seems we need to fix things upstream here. Could you please confirm or deliver a short description about how this shoukd be fixed. Woukd like to understabd early enough if we have external dependencies and I'm not sure to fully understand the implications of @Markus-Rost analysis.

**Comment by Markus-Rost** [comment]
[MathJax](https://www.mathjax.org/) is a JavaScript library, meaning this is blocked by us not including JS in any case (#2310).

There are three different wiki extensions using MathJax:

- [Extension:MathJax](https://www.mediawiki.org/wiki/Extension:MathJax) - used by proofwiki.org and encyclopediaofmath.org
- [Extension:SimpleMathJax](https://www.mediawiki.org/wiki/Extension:SimpleMathJax) - used by terraria.wiki.gg
- [Extension:Math](https://www.mediawiki.org/wiki/Extension:Math) - only with mode `mathjax`

Even with #2483 implemented, none of these extensions would work due to the way they are loading the MathJax library.
We would likely have to hardcode support for the different MathJax extensions, and even that will be difficult due to the different MathJax config options the wikis can have.

From a mwoffliner perspective, the implementation of Extension:Math would be best for a ZIM, due to being the only one to return [MathML](https://developer.mozilla.org/en-US/docs/Web/MathML) for no-Js users before running MathJax for compatibility with older browsers. Meaning even without support for MathJax, we would have working math formula on all browsers since 2023.

**Comment by Markus-Rost** [comment]
@kelson42 the only upstream part of this issue would be to request the extensions to use a different hook for adding the JS ResourceLoader module, so that we can see it in the API output. Do note however, that Extension:MathJax is currently unmaintained.

But as mentioned in my previous comment, even with that being done it will be difficult to properly support MathJax at all.

**Comment by kelson42** [comment]
@Markus-Rost Thank you for your patience explaining this. Trying to think out of the box: what solution could be implemented on our side to transform the mathjax in natively supported mathml? Or why we don't do a special piece of code to detect if mathjax is required and insert the javascript if "yes", independently to any other MediaWiki related concept?

**Comment by Markus-Rost** [comment]
> what solution could be implemented on our side to transform the mathjax in natively supported mathml? 

Any solution for that will likely be more difficult than just supporting MathJax itself.

> Or why we don't do a special piece of code to detect if mathjax is required and insert the javascript if "yes", independently to any other MediaWiki related concept?

The special piece of code to detect if mathjax is required would be the hardcoded support for the MathJax extensions I mentioned in my comment. The formula for MathJax are plain text inside the html, so there are no special elements we could use for detection and just looking for the plain text markers will create a lot of false positives on other wikis.

**Comment by kelson42** [comment]
> The special piece of code to detect if mathjax is required would be the hardcoded support for the MathJax extensions I mentioned in my comment.

Can we just not look if the Mediawiki has the Mathjax extension and include the piece of js on all page if "yes".

**Comment by Jaifroid** [comment]
> Or why we don't do a special piece of code to detect if mathjax is required and insert the javascript if "yes", independently to any other MediaWiki related concept?

My experience with this might be interesting. Back in the days when the KJS engine would crash trying to decode hundreds of SVG representations of equations on Wikipedia maths pages like `Sine`, I added support (in the then Windows UWP app) for stripping the SVGs and interpreting the MathML directly with MathJax as a dependency of the reader, and then later its faster cousin [KaTeX](https://katex.org/), injected into pages where MathML is present. This was way faster and made such articles readable. It also worked in Mathoverflow before the ZIM included the necessary JS modules.

That reader-provided support is still in the PWA, and will still kick in in some configurations. To my surprise it kicks in with ProofWiki, albeit not all equations are properly converted (because the configuration is for the symbols Wikipedia uses, which doesn't include special extensions):

<img width="1591" height="856" alt="Image" src="https://github.com/user-attachments/assets/13666480-2d8a-4731-901e-cb300732a3c0" />

My point is not that this should be done reader-side, but that adding MathJax or KaTeX modules into the page as a dependency of mwOffliner is not in principle a hard configuration. My implementation gets out of the way if it sees MathJax or KaTeX already loaded by the content. But of course, as @Markus-Rost says, a challenge for a universal solution in mwOffliner is adapting to the differet flavours of MathJax needed in each case.

**Comment by benoit74** [comment]
Removing the `upstream` label since we've decided we will solve this "on our own" in the scraper.

Thinking a bit about this, I think I will experiment with a new CLI parameter `--mathjax-zip-url` to specify the URL where one should download MathJax code to add to the ZIM and `--mathjax-config-url` to specify the URL where one should download MathJax configuration for the given ZIM (to be added to the ZIM as well).

Looks like proofwiki uses a very large config with custom macros (among which the `\struct` @Jaifroid spotted) while encyclopediaofmath.org seems to use no configuration at all.

I feel like this would be the most versatile solution to support MathJax in cases where we need it and avoid having to reimplement many edge cases. 

It pushes lots of responsibilities towards mwoffliner user which I don't really like, but I do feel like this is the solution which alleviates much pressure on mwoffliner maintenance. This could allow us to gain experience on MathJax usage and configuration in the wild, and then consider if we can automate (some) scenario.

Not sure we will merge this straight-away, but at least it will help better understand the problem at hand.

@Markus-Rost do you have any idea of a sample article where terraria.wiki.gg does uses MathJax? Or a tool/API to find this?
